src/main/java/org/springframework/data/elasticsearch/core/DocumentAdapters.java

```diff
@@ -0,0 +1,674 @@
+		 * @see java.util.Map#put(java.lang.Object, java.lang.Object)
+		 */
+		@Override
+		public Object put(String key, Object value) {
+			throw new UnsupportedOperationException();
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see java.util.Map#remove(java.lang.Object)
+		 */
+		@Override
+		public Object remove(Object key) {
+			throw new UnsupportedOperationException();
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see java.util.Map#putAll(Map)
+		 */
+		@Override
+		public void putAll(Map<? extends String, ?> m) {
+			throw new UnsupportedOperationException();
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see java.util.Map#clear()
+		 */
+		@Override
+		public void clear() {
+			throw new UnsupportedOperationException();
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see java.util.Map#keySet()
+		 */
+		@Override
+		public Set<String> keySet() {
+			return documentFields.stream().map(DocumentField::getName).collect(Collectors.toCollection(LinkedHashSet::new));
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see java.util.Map#values()
+		 */
+		@Override
+		public Collection<Object> values() {
+			return documentFields.stream().map(DocumentFieldAdapter::getValue).collect(Collectors.toList());
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see java.util.Map#entrySet()
+		 */
+		@Override
+		public Set<Entry<String, Object>> entrySet() {
+			return documentFields.stream().collect(Collectors.toMap(DocumentField::getName, DocumentFieldAdapter::getValue))
+					.entrySet();
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see java.util.Map#forEach(java.util.function.BiConsumer)
+		 */
+		@Override
+		public void forEach(BiConsumer<? super String, ? super Object> action) {
+
+			Objects.requireNonNull(action);
+
+			documentFields.forEach(field -> {
+				action.accept(field.getName(), getValue(field));
+			});
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.data.elasticsearch.Document#toJson()
+		 */
+		@Override
+		public String toJson() {
+
+			JsonFactory nodeFactory = new JsonFactory();
+			try {
+				ByteArrayOutputStream stream = new ByteArrayOutputStream();
+				JsonGenerator generator = nodeFactory.createGenerator(stream, JsonEncoding.UTF8);
+				generator.writeStartObject();
+				for (DocumentField value : documentFields) {
+					if (value.getValues().size() > 1) {
+						generator.writeArrayFieldStart(value.getName());
+						for (Object val : value.getValues()) {
+							generator.writeObject(val);
+						}
+						generator.writeEndArray();
+					} else {
+						generator.writeObjectField(value.getName(), value.getValue());
+					}
+				}
+				generator.writeEndObject();
+				generator.flush();
+				return new String(stream.toByteArray(), StandardCharsets.UTF_8);
+			} catch (IOException e) {
+				throw new ElasticsearchException("Cannot render JSON", e);
+			}
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see java.lang.Object#toString()
+		 */
+		@Override
+		public String toString() {
+			return getClass().getSimpleName() + "@" + this.id + "#" + this.version + " " + toJson();
```

Comment:
```suggestion
			return getClass().getSimpleName() + '@' + this.id + '#' + this.version + ' ' + toJson();
```
IntelliJ thinks this is better